Bug 1402528 - recipe-client-addon: Watch prefs and enable/disable as needed.

mythmon opened the pull request:
This is a fairly large (in terms of lines) change to `RecipeRunner`'s startup and pref handling. I think it should be fairly straight forward though. It is mostly just pretty verbose.

Osmose commented on recipe-client-addon/test/browser/browser_RecipeRunner.js:
```diff
@@ -356,22 +358,80 @@ decorate_task(
+  withStub(CleanupManager, "addCleanupHandler"),
+
+  async function testPrefWatching(runStub, enableStub, disableStub, addCleanupHandlerStub) {
+    await RecipeRunner.init();
```
Could we potentially have multiple watchers registered during this test (since this was already called at the start of the session)? Feels like that could cause some hard-to-trace test errors down the road.

mythmon replied:
Since `enable` is stubbed out, and the watchers are set through that, no. I think we're ok here.

rhelmer commented on recipe-client-addon/lib/RecipeRunner.jsm:
```diff
@@ -39,88 +39,144 @@ Cu.importGlobalProperties(["fetch"]);
 
+  registerTimer() {
+    this.updateRunInterval();
+    CleanupManager.addCleanupHandler(() => timerManager.unregisterTimer(TIMER_NAME));
```
Shouldn't throw, just logs and moves on:
http://searchfox.org/mozilla-central/rev/e62604a97286f49993eb29c493672c17c7398da9/toolkit/components/timermanager/nsUpdateTimerManager.js#345

rhelmer commented on recipe-client-addon/lib/RecipeRunner.jsm:
```diff
@@ -39,88 +39,144 @@ Cu.importGlobalProperties(["fetch"]);
+    }
+  },
 
+  enable() {
```
Current code handles `registerTimer` being called multiple times:
http://searchfox.org/mozilla-central/rev/e62604a97286f49993eb29c493672c17c7398da9/toolkit/components/timermanager/nsUpdateTimerManager.js#319

Seems cleaner not to do it though 🎨

mythmon commented:
I updated enable and disable to check the current state before acting, which should fix the idempotency issues brought up in review.

mythmon commented on recipe-client-addon/lib/RecipeRunner.jsm:
```diff
@@ -39,88 +39,144 @@ Cu.importGlobalProperties(["fetch"]);
-    if (!this.checkPrefs()) {
-      return;
-    }
+    this.enabled = null;
```
I think the changes we want to enable and disable below necessitates keeping this bit of state. Sorry 'bout your mantra.

Osmose commented on recipe-client-addon/lib/RecipeRunner.jsm:
```diff
@@ -39,88 +39,152 @@ Cu.importGlobalProperties(["fetch"]);
+    if (this.enabled) {
+      return;
+    }
+    this.enabled = true;
```
Minor: We should update this at the end of the function in case the changes throw and we're not actually enabled.

Osmose replied:
We could probably use a test that multiple calls to enable/disable do not register multiple watchers.

Osmose commented on recipe-client-addon/lib/RecipeRunner.jsm:
```diff
@@ -39,88 +39,152 @@ Cu.importGlobalProperties(["fetch"]);
+
+    // Only run if Unified Telemetry is enabled.
+    if (!Services.prefs.getBoolPref(TELEMETRY_ENABLED_PREF)) {
+      log.debug("Disabling RecipeRunner because Unified Telemetry is disabled.");
```
I like that this states what it's doing, can we update the other logs to also state that they're disabling Shield?

Osmose replied:
I wonder if having multiple log messages claiming to be disabling the recipe runner would be confusing to, say, QA. Dunno if it's better to disable immediately and return, or list out the reasons like it is currently.

mythmon commented:
Updated and rebased on master to fix merge conflicts.

mythmon commented:
I think my most recent fixes broke the TC failure. I'll look into that on Monday.

mythmon commented:
I couldn't reproduce the test failures locally. I re-pushed to trigger TC again. Maybe it was some sort of intermittent?

mythmon commented:
It doesn't seem like it was intermittent, but I don't really know what could have happened. @rhelmer do you have any ideas?

rhelmer commented on recipe-client-addon/lib/RecipeRunner.jsm:
```diff
@@ -39,88 +39,147 @@ Cu.importGlobalProperties(["fetch"]);
-    // Watch for the run interval to change, and re-register the timer with the new value
-    prefs.addObserver(RUN_INTERVAL_PREF, this);
-    CleanupManager.addCleanupHandler(() => prefs.removeObserver(RUN_INTERVAL_PREF, this));
+    CleanupManager.addCleanupHandler(this.unwatchPrefs.bind(this));
```
Hm. Are you sure this works? I thought this would give you a different reference, when you need the same for both add/remove... I've always done this like:

```
this.unwatchPrefs = this.unwatchPrefs.bind(this);
addListener(this.unwatchPrefs);
/* later... */
removeListener(this.unwatchPrefs);
```

mythmon replied:
You're right that if we needed to remove a listener later, this would be a problem. But we never remove cleanup handlers, so we won't need this reference again. CleanupHandler is an Shield-ism. this will be the same in the callback (unwatchPrefs), soServices.prefs.removeObserver(pref, this)` (line 107) will work fine.

mythmon commented on recipe-client-addon/lib/RecipeRunner.jsm:
```diff
@@ -39,88 +39,147 @@ Cu.importGlobalProperties(["fetch"]);
 this.EXPORTED_SYMBOLS = ["RecipeRunner"];
 
 const log = LogManager.getLogger("recipe-runner");
-const prefs = Services.prefs.getBranch("extensions.shield-recipe-client.");
```
I did this because we are accessing prefs from more than just the `extensions.shield-recipe.client` sub-tree. It seemed awkward to use `prefs.getCharPref` in someplaces and `Services.prefs.getCharPref` in others, depending on which preference was being accessed.

mythmon commented:
Doh. I forgot to update the preference name in the tests when I changed it in the code. I've also rebased this on master to pick up the TC fixes from #1110.

@rhelmer this should be ready for a final review, and if you approve it, it should be ready to merge.
